File: archived/DateTimeFormatting/js/js/scenario3-parameterizedTemplate.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var DateTimeFormatting = Windows.Globalization.DateTimeFormatting;
    var DateTimeFormatter = DateTimeFormatting.DateTimeFormatter;

    var page = WinJS.UI.Pages.define("/html/scenario3-parameterizedTemplate.html", {
        ready: function (element, options) {
            document.getElementById("displayButton").addEventListener("click", doDisplay, false);
        }
    });

    function doDisplay() {
        // This scenario uses the DateTimeFormatter class
        // to format a date/time by specifying a template via parameters.  Note that the current language
        // and region value will determine the pattern of the date returned based on the
        // specified parts.

        SdkSample.clearFormattingResults();
        SdkSample.appendCurrentLanguage();

        // Obtain the date that will be formatted.
        var dateToFormat = new Date();

        SdkSample.appendFormattingResults("Dates:", [
            new DateTimeFormatter(
                DateTimeFormatting.YearFormat.full,
                DateTimeFormatting.MonthFormat.abbreviated,
                DateTimeFormatting.DayFormat.default,
                DateTimeFormatting.DayOfWeekFormat.abbreviated),

            new DateTimeFormatter(
                DateTimeFormatting.YearFormat.abbreviated,
                DateTimeFormatting.MonthFormat.abbreviated,
                DateTimeFormatting.DayFormat.default,
                DateTimeFormatting.DayOfWeekFormat.none),

            new DateTimeFormatter(
                DateTimeFormatting.YearFormat.full,
                DateTimeFormatting.MonthFormat.full,
                DateTimeFormatting.DayFormat.none,
                DateTimeFormatting.DayOfWeekFormat.none),

            new DateTimeFormatter(
                DateTimeFormatting.YearFormat.none,
                DateTimeFormatting.MonthFormat.full,
                DateTimeFormatting.DayFormat.default,
                DateTimeFormatting.DayOfWeekFormat.none)
        ], function (formatter) {
            // Perform the actual formatting.
            return formatter.template + ": " + formatter.format(dateToFormat);
        });

        SdkSample.appendFormattingResults("Times:", [
            new DateTimeFormatter(
                DateTimeFormatting.HourFormat.default,
                DateTimeFormatting.MinuteFormat.default,
                DateTimeFormatting.SecondFormat.default),

            new DateTimeFormatter(
                DateTimeFormatting.HourFormat.default,
                DateTimeFormatting.MinuteFormat.default,
                DateTimeFormatting.SecondFormat.none),

            new DateTimeFormatter(
                DateTimeFormatting.HourFormat.default,
                DateTimeFormatting.MinuteFormat.none,
                DateTimeFormatting.SecondFormat.none)
        ], function (formatter) {
            // Perform the actual formatting.
            return formatter.template + ": " + formatter.format(dateToFormat);
        });
    }
})();

File: Samples/DateTimeFormatting/cppwinrt/Scenario3_ParameterizedTemplate.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario3_ParameterizedTemplate.h"
#include "Scenario3_ParameterizedTemplate.g.cpp"

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::Globalization;
using namespace Windows::Globalization::DateTimeFormatting;
using namespace Windows::UI::Xaml;

namespace
{
    void GenerateResult(std::wostream& results, DateTimeFormatter const& formatter, DateTime dateToFormat)
    {
        // Perform the actual formatting. 
        results << std::wstring_view(formatter.Template()) << L": " << std::wstring_view(formatter.Format(dateToFormat)) << std::endl;
    }
}

namespace winrt::SDKTemplate::implementation
{
    Scenario3_ParameterizedTemplate::Scenario3_ParameterizedTemplate()
    {
        InitializeComponent();
    }

    void Scenario3_ParameterizedTemplate::Display_Click(IInspectable const&, RoutedEventArgs const&)
    {
        // This scenario uses the Windows.Globalization.DateTimeFormatting.DateTimeFormatter class
        // to format a date/time by specifying a template via parameters.  Note that the current application language 
        // language and region value will determine the pattern of the date returned based on the 
        // specified parts.

        //Get the current application context language
        hstring currentLanguage = ApplicationLanguages::Languages().GetAt(0);

        // Formatters for dates
        std::array dateFormatters{
            DateTimeFormatter(
                YearFormat::Full,
                MonthFormat::Abbreviated,
                DayFormat::Default,
                DayOfWeekFormat::Abbreviated),

            DateTimeFormatter(
                YearFormat::Abbreviated,
                MonthFormat::Abbreviated,
                DayFormat::Default,
                DayOfWeekFormat::None),

            DateTimeFormatter(
                YearFormat::Full,
                MonthFormat::Full,
                DayFormat::None,
                DayOfWeekFormat::None),

            DateTimeFormatter(
                YearFormat::None,
                MonthFormat::Full,
                DayFormat::Default,
                DayOfWeekFormat::None),
        };

        // Formatters for times
        std::array timeFormatters{
            DateTimeFormatter(
                HourFormat::Default,
                MinuteFormat::Default,
                SecondFormat::Default),

            DateTimeFormatter(
                HourFormat::Default,
                MinuteFormat::Default,
                SecondFormat::None),

            DateTimeFormatter(
                HourFormat::Default,
                MinuteFormat::None,
                SecondFormat::None),
        };

        // Obtain the date that will be formatted.
        DateTime dateToFormat = clock::now();

        // Keep the results here
        std::wostringstream results;
        results << L"Current application context language: " << std::wstring_view(currentLanguage) << std::endl << std::endl;
        results << L"Dates:" << std::endl;

        // Generate the results.
        for (auto&& formatter : dateFormatters)
        {
            GenerateResult(results, formatter, dateToFormat);
        }

        results << std::endl;
        results << L"Times:" << std::endl;

        // Generate the results.
        for (auto&& formatter : timeFormatters)
        {
            GenerateResult(results, formatter, dateToFormat);
        }

        // Display the results.
        OutputTextBlock().Text(results.str());
    }
}
